src/cljam/io/bam_index/writer.clj

```diff
@@ -310,7 +228,114 @@
          make-index-fn
          (finalize-index nrefs))))
 
+;; Public
+;; ------
+
+;; Writing index
+;; -----------
+
+(defn write-index*!
+  [wtr ^long nrefs indices]
+  ;; magic
+  (lsb/write-bytes wtr (.getBytes ^String bai-magic))
+  ;; n_ref
+  (lsb/write-int wtr nrefs)
+  (dotimes [i nrefs]
+    (let [index (get indices i)
+          n-bin (count (:bin-index index))]
+        ;; bins
+      (if (zero? n-bin)
+        (lsb/write-int wtr 0)
+        (do
+            ;; # of bins
+          (lsb/write-int wtr (inc n-bin))
+          (doseq [bin (:bin-index index)]
+            (write-bin wtr (:bin bin) (:chunks bin)))
+            ;; meta data
+          (write-meta-data wtr (:meta-data index))))
+        ;; linear index
+      (lsb/write-int wtr (count (:linear-index index)))
+      (doseq [l (:linear-index index)]
+        (lsb/write-long wtr l))))
+    ;; no coordinate alignments
+  (lsb/write-long wtr (:no-coordinate-alns indices)))
+
+;; ### Making index
+
+(defn make-index*
+  "Calculates index from the references and alignments, returning it as a map.
+  Returned index is still intermediate. It must be passed to finalize function
+  in the final stage."
+  [alns]
+  (loop [[^BAMPointerBlock aln & rest] alns
+         rid (.ref-id aln)
+         idx-status (init-index-status)
+         no-coordinate-alns 0
+         indices {}]
+    (if aln
+      (let [rid' (.ref-id aln)
+            new-ref? (not= rid' rid)
+            idx-status' (update-index-status
+                         (if new-ref? (init-index-status) idx-status) aln)
+            no-coordinate-alns' (if (zero? (.pos aln))
+                                  (inc no-coordinate-alns)
+                                  no-coordinate-alns)
+            indices' (if new-ref?
+                       (assoc indices rid idx-status)
+                       indices)]
+        (recur rest rid' idx-status' no-coordinate-alns' indices'))
+      (assoc indices rid idx-status
+             :no-coordinate-alns no-coordinate-alns))))
+
+(defn update-last-pointer [index eof-ptr]
```

Comment:
Add a docstring

Reply:
And would you add a unit test for this function?

Reply:
Docstring added.
Test added.